File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/typography/wrapping/AutoHyphenation.swift
//
//  AutoHyphenation.swift
//  StyleEngine/typography/wrapping — wave 40 (lane T2).
//
//  css-text-3 §6.1 `hyphens: auto`, the DICTIONARY half — the wall
//  HyphensApplier and SoftHyphenPolicy have both been naming since
//  wave 37, now taken down.
//
//  ── WHAT THE OLD WALL ACTUALLY WAS ──────────────────────────────────
//  Two separate claims were being made, and only ONE of them was true:
//
//   (1) "the IR carries no language, so no dictionary can be selected".
//       TRUE until wave 37 (lane W4) shipped THE LANG WIRE — `meta.lang`,
//       the element's COMPUTED content language, already resolved by the
//       producer through HTML §3.2.6.2's own-lang → nearest-ancestor →
//       `<html>` ladder (schema/spec/04-metadata-fields.md). It is on the
//       wire, decoded by IRWireV2Reader, and was simply never read by
//       this runtime.
//
//   (2) "TextKit hyphenation needs NSParagraphStyle on a UIKit label,
//       and ImageRenderer refuses to rasterise platform views".
//       TRUE of `NSAttributedString.hyphenationFactor` and of
//       `usesDefaultHyphenation` — but those were never the only seam.
//       CoreFoundation exposes the hyphenation DICTIONARY itself as a
//       plain string query: `CFStringIsHyphenationAvailableForLocale`
//       + `CFStringGetHyphenationLocationBeforeIndex`, available since
//       iOS 4.2, no view, no layout manager, no ImageRenderer exposure.
//       Ask it where a word may break and the answer is a set of
//       offsets — exactly the input this platform's line breaker needs,
//       because on iOS the LINE BREAKER IS OURS (GreedyLineBreaker
//       pre-breaks every wrappable run with hard newlines; see its
//       banner for why TextKit's own push-out strategy forced that).
//
//  So the ImageRenderer wall is real for TextKit's hyphenator and
//  IRRELEVANT to CoreFoundation's dictionary. That is the whole finding.
//
//  ── THE GATE ────────────────────────────────────────────────────────
//  `engaged` requires BOTH `hyphens: auto` and a non-blank language tag,
//  which is §6.1's own condition ("a hyphenation resource appropriate to
//  the LANGUAGE of the text") and the one WPT css-text/hyphens-auto-001
//  asserts the contrapositive of: "automatic hyphenation must not work
//  without language tagging". `meta.lang` exists only on IR extracted
//  from real HTML — every hand-authored fixture is a CSS envelope with
//  no `_lang` key — so this file cannot move a committed baseline.
//
//  Twin: `AutoHyphenation.kt` (Compose) carries the identical three pure
//  entry points. It has no hyphenator section because Android's breaker
//  is INSIDE `Text`: there the switch is `TextStyle.hyphens = .Auto` +
//  a `LocaleList`, and Minikin consults its own `/system/usr/hyphen-data`
//  patterns. Same decision, different owner of the break.
//

import Foundation

enum AutoHyphenation {

    // MARK: - The decision (byte-parallel with AutoHyphenation.kt)

    /// Does this run ask for — and qualify for — dictionary hyphenation?
    ///
    /// - Parameters:
    ///   - hyphensMode: resolved `hyphens` keyword in either spelling the
    ///     two extractors produce (the wire authors upper-case `"AUTO"`,
    ///     this runtime's HyphensExtractor lowercases).
    ///   - lang: the run's `meta.lang`. nil/blank = the document declares
    ///     no language, which §6.1 leaves with no resource to select, so
    ///     `auto` degrades to `manual`'s explicit opportunities exactly as
    ///     before this file existed.
    static func engaged(_ hyphensMode: String?, _ lang: String?) -> Bool {
        guard hyphensMode?.lowercased() == "auto", let l = lang else { return false }
        return !l.trimmingCharacters(in: .whitespaces).isEmpty
    }

    /// The BCP-47 tag to hand the hyphenator, or nil when not engaged.
    /// Returned VERBATIM apart from surrounding whitespace: BCP-47
    /// matching is case-insensitive and subtag-truncating (RFC 4647
    /// §2.1/§3.4) and `CFLocale` canonicalises at construction, so
    /// normalising here would only lose the wire's round-trip.
    static func localeTag(_ hyphensMode: String?, _ lang: String?) -> String? {
        guard engaged(hyphensMode, lang) else { return nil }
        return lang?.trimmingCharacters(in: .whitespaces)
    }

    /// css-text-4 §6.2 `hyphenate-limit-chars: auto` — the minimum word
    /// length any engine in this comparison will hyphenate (Chromium 2+2,
    /// Minikin MIN_PREFIX 2 / MIN_SUFFIX 3; five is the shared floor).
    private static let minHyphenatableWord = 5

    /// Could a dictionary find a break inside `text`? The hyphenation
    /// half of "is this line breakable?", the companion to
    /// `DecorationOps.hasSoftWrapOpportunity`'s UAX #14 half. A digit run
    /// (`00000`) has no letters and therefore nowhere to hyphenate, so a
    /// line made of one still counts as unbreakable and keeps the §5.2
    /// overflow rule — the Compose twin measured that on
    /// css-text/hyphens-punctuation-001.
    static func hasDictionaryOpportunity(_ text: String) -> Bool {
        var run = 0
        for c in text {
            if c.isLetter {
                run += 1
                if run >= minHyphenatableWord { return true }
            } else {
                run = 0
            }
        }
        return false
    }

    // MARK: - The hyphenator (the half Minikin owns on Android)

    /// Cache of resolved locales: `CFLocale` construction plus the
    /// availability probe is not free and a corpus run asks for the same
    /// two or three tags thousands of times. Keyed by the VERBATIM tag.
    /// nil value = "this tag has no dictionary on this device" (a real,
    /// cacheable answer — never retried).
    private static var localeCache: [String: CFLocale?] = [:]

    /// The `CFLocale` for `tag` if — and only if — the device ships a
    /// hyphenation dictionary for it. `CFStringIsHyphenationAvailableForLocale`
    /// is the honest probe: iOS bundles patterns for a fixed language set,
    /// and asking for a language it lacks must degrade to `manual`, not
    /// silently produce no breaks with the switch reported as "on".
    static func locale(for tag: String) -> CFLocale? {
        if let cached = localeCache[tag] { return cached }
        // Built through NSLocale rather than `CFLocaleCreate`: CFLocale and
        // NSLocale are toll-free bridged, and Swift's importer maps
        // `CFLocaleIdentifier` to a type no String or CFString will coerce
        // to — so the ObjC initialiser is the only spelling that compiles
        // for a runtime-supplied BCP-47 tag.
        let l = NSLocale(localeIdentifier: tag) as CFLocale
        let usable: CFLocale? =
            CFStringIsHyphenationAvailableForLocale(l) ? l : nil
        localeCache[tag] = usable
        return usable
    }

    /// Every hyphenation opportunity inside `word`, as ASCENDING UTF-16
    /// offsets: a hyphen may be inserted BEFORE the character at each
    /// offset (`CFStringGetHyphenationLocationBeforeIndex`'s own contract).
    ///
    /// The API answers one point at a time, searching BACKWARDS from a
    /// given index, so the enumeration walks down from the end and
    /// reverses. Empty for a word the dictionary does not break (and for
    /// anything shorter than the §6.2 floor, which CF also refuses).
    static func breakOffsets(in word: String, locale: CFLocale) -> [Int] {
        let s = word as CFString
        let length = CFStringGetLength(s)
        guard length >= minHyphenatableWord else { return [] }
        let range = CFRange(location: 0, length: length)
        var out: [Int] = []
        var searchBefore = length
        // Bounded by construction: every accepted index is strictly
        // smaller than the previous `searchBefore`, so the walk descends.
        while searchBefore > 0 {
            let idx = CFStringGetHyphenationLocationBeforeIndex(
                s, searchBefore, range, 0, locale, nil)
            guard idx != kCFNotFound, idx > 0, idx < searchBefore else { break }
            out.append(idx)
            searchBefore = idx
        }
        return out.reversed()
    }

    /// The hyphen character this locale's dictionary asks the UA to paint
    /// (§6.1 leaves it UA-defined; css-text-4's `hyphenate-character`
    /// overrides it, and a declared value wins at the call site).
    /// U+2010 HYPHEN is what CoreFoundation reports for the Latin
    /// languages and what Chromium paints, so it is also the fallback.
    static let defaultHyphenCharacter = "\u{2010}"
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/typography/wrapping/AutoHyphenation.kt
package com.styleconverter.runtime.typography.wrapping

/**
 * AutoHyphenation.kt
 * typography/wrapping — wave 40 (lane T2).
 *
 * css-text-3 §5.3 `hyphens: auto`, the DICTIONARY half — the switch that
 * wave 37 named and wave 39's A3 finding measured as still off.
 *
 * ── WHAT WAS MISSING ────────────────────────────────────────────────
 * [com.styleconverter.runtime.typography.TextWrapApplier.getHyphens]
 * has mapped `HyphensMode.AUTO → Hyphens.Auto` since Phase 6, but
 * nothing ever read it: the renderer builds its own `TextStyle` and
 * never set `hyphens`, so every Compose run laid out at Compose's
 * default `Hyphens.None` (Minikin `hyphenationFrequency = NONE`). The
 * mapping was a function with no call site. Measured cost, wave39-final
 * android-ref: css-text/hyphens-auto-010 0.8530, hyphens-span-002
 * 0.8842, hyphens-punctuation-001 0.9385, hyphens-out-of-flow-002
 * 0.8782 — in every one the Chromium ref hyphenates ("regu-/lation",
 * "high-/way", "exam-/ple") and Android renders the whole word.
 *
 * ── WHY IT COULD NOT BE SWITCHED ON BEFORE, AND CAN NOW ─────────────
 * §5.3 makes the hyphenation resource LANGUAGE-dependent, and WPT
 * `css-text/hyphens/hyphens-auto-001` asserts the contrapositive in so
 * many words: "automatic hyphenation must not work without language
 * tagging" (its own `<div>` carries `hyphens: auto` and NO `lang`, and
 * both natives score 0.99 today precisely because they do not
 * hyphenate). Until wave 37 the IR had no language channel, so the
 * runtime could not tell a tagged run from an untagged one and took the
 * untagged reading for both — the honest choice, and the one recorded
 * in [SoftHyphenPolicy.wantsDictionaryHyphenation]'s banner.
 *
 * Wave 37 (lane W4) shipped THE LANG WIRE: `meta.lang`, the element's
 * COMPUTED content language, already resolved by the producer through
 * HTML §3.2.6.2's own-lang → nearest-ancestor → `<html>` ladder
 * (schema/spec/04-metadata-fields.md). That is exactly the missing
 * input, so the two readings can finally be separated.
 *
 * ── THE GATE, AND WHY IT IS INERT FOR EVERY COMMITTED BASELINE ──────
 * [engaged] is true only when BOTH halves hold: the resolved keyword is
 * `auto` AND the run carries a non-blank language tag. `meta.lang`
 * exists only on IR produced from real HTML (tools/titan's WPT
 * extractor sets `_lang`; CssParsing forwards it verbatim) — every
 * hand-authored fixture under `fixtures/properties/` and
 * `fixtures/components/` is a CSS envelope with no `_lang` key, so
 * `lang` is null there and this file cannot change a pixel of the 363
 * committed `tools/visual/baseline/` captures. That is a stronger,
 * spec-derived gate than a WPT-mode flag would be: it is the same
 * condition §5.3 itself imposes.
 *
 * Pure Kotlin (no Compose, no Android) so the JVM suite pins the
 * decision without a device. The Compose-typed half — `Hyphens.Auto` +
 * the `LocaleList` Minikin selects the dictionary with — lives in
 * TextWrapApplier, which already owns the `hyphens` mapping.
 *
 * Twin: `AutoHyphenation.swift` (SwiftUI), same two entry points and
 * the same gate; the platforms differ only in the hyphenator they hand
 * the decision to (Minikin here, CoreFoundation's
 * `CFStringGetHyphenationLocationBeforeIndex` there).
 */
object AutoHyphenation {

    /**
     * Does this run ask for — and qualify for — dictionary hyphenation?
     *
     * @param hyphensMode the resolved `hyphens` keyword, in either
     *   spelling the two extractors produce (the wire authors upper-case
     *   `"AUTO"`, iOS's extractor lowercases) — compared case-insensitively.
     * @param lang the run's COMPUTED content language (`meta.lang`).
     *   Null/blank means the document declares none, which §5.3 leaves
     *   with no resource to select: `auto` then degrades to `manual`'s
     *   explicit opportunities, exactly as before this file existed.
     */
    @JvmStatic
    fun engaged(hyphensMode: String?, lang: String?): Boolean =
        hyphensMode?.lowercase() == "auto" && !lang.isNullOrBlank()

    /**
     * The BCP-47 tag to hand the platform hyphenator, or null when
     * [engaged] is false.
     *
     * Returned VERBATIM apart from surrounding whitespace: BCP-47
     * matching is case-insensitive and subtag-truncating (RFC 4647
     * §2.1/§3.4), Minikin lowercases at lookup, and canonicalising here
     * would only lose the round-trip the wire deliberately preserves
     * (`meta.lang` keeps `eN-Us` as authored — see the schema note).
     * `en-us` therefore selects the same `hyph-en-us` pattern set as
     * `en-US`, and a bare `en` truncates to the same family.
     */
    @JvmStatic
    fun localeTag(hyphensMode: String?, lang: String?): String? =
        if (engaged(hyphensMode, lang)) lang!!.trim() else null

    /**
     * css-text-4 §6.3.4 `hyphenate-limit-chars: auto` — the MINIMUM word
     * length a UA will hyphenate at all. Chromium (`kMinimumPrefixLength`
     * 2 + `kMinimumSuffixLength` 2) and Minikin (`Hyphenator` MIN_PREFIX 2
     * / MIN_SUFFIX 3) both refuse anything shorter, so five is the floor
     * every engine in this repo's comparison agrees on.
     */
    private const val MIN_HYPHENATABLE_WORD = 5

    /**
     * Could a dictionary find a break inside [text]?
     *
     * This is the HYPHENATION half of the "is this line breakable?"
     * question that css-text-3 §5.5's rule B turns on
     * ([GreedyLineBreaker.hasUnbreakableOverflowingLine] owns the UAX #14
     * half through `DecorationOps.hasSoftWrapOpportunity`). Rule B exists
     * to stop the platform emergency-breaking a word that has NOWHERE to
     * break; once `hyphens: auto` is engaged a long word usually DOES have
     * somewhere, and letting rule B fire on it would pre-break the whole
     * run with `softWrap = false` — suppressing the hyphenation the same
     * wave just switched on.
     *
     * But the converse still matters, which is why this is not simply
     * `true`: a hyphenator only breaks WORDS. Measured on
     * css-text/hyphens-punctuation-001, whose `width: 5ch` boxes hold
     * `00000 example 00000` — the digit runs overflow with no dictionary
     * point anywhere in them, so rule B must still claim that run or
     * Minikin desperate-breaks `00000` into `0000`/`0` (android-ref
     * 0.9385 → 0.8971 with a blanket veto, restored by this predicate).
     *
     * Deliberately a LETTER-RUN test and nothing cleverer: the pattern
     * files are letter-only, the runtime cannot consult them from pure
     * Kotlin, and the only alternative — asking the platform to lay the
     * word out — is exactly the work rule B is trying to decide about.
     * A ≥5-letter run that the dictionary happens NOT to break is the one
     * false positive; it costs a desperate break on a word we would
     * otherwise have overflowed, which is the pre-wave-38 behaviour.
     */
    @JvmStatic
    fun hasDictionaryOpportunity(text: String): Boolean {
        var run = 0
        for (c in text) {
            if (c.isLetter()) {
                run++
                if (run >= MIN_HYPHENATABLE_WORD) return true
            } else {
                run = 0
            }
        }
        return false
    }
}
